logic/Rule.java

```diff
@@ -258,27 +256,30 @@ public Value asValue() {
             throw GraknException.of(INVALID_CASTING, className(this.getClass()), className(Value.class));
         }
 
-        private void copyAdditionalConstraints(Set<Variable> fromVars, Set<Variable> toVars) {
-            Map<Variable, Variable> nonAnonFromVarsMap = fromVars.stream()
-                    .filter(variable -> !variable.id().reference().isAnonymous())
-                    .collect(Collectors.toMap(e -> e, e -> e)); // Create a map for efficient lookups
-            toVars.stream().filter(variable -> !variable.id().reference().isAnonymous())
-                    .forEach(copyTo -> {
-                        if (nonAnonFromVarsMap.containsKey(copyTo)) {
-                            Variable copyFrom = nonAnonFromVarsMap.get(copyTo);
-                            if (copyTo.isThing() && copyFrom.isThing()) {
-                                ConstraintCopier.copyIsaAndValues(copyFrom.asThing(), copyTo.asThing());
-                            } else if (copyTo.isType() && copyFrom.isType()) {
-                                ConstraintCopier.copyLabelSubAndValueType(copyFrom.asType(), copyTo.asType());
-                            } else throw GraknException.of(ILLEGAL_STATE);
-                        }
-                    });
+        public Has.Variable asVariableHas() {
+            throw GraknException.of(INVALID_CASTING, className(this.getClass()), className(Has.Variable.class));
         }
 
-        public static class Relation extends Conclusion<RelationConstraint> {
+        public Has.Explicit asExplicitHas() {
+            throw GraknException.of(INVALID_CASTING, className(this.getClass()), className(Has.Explicit.class));
+        }
 
-            public Relation(RelationConstraint constraint, Set<Variable> whenContext) {
-                super(constraint, whenContext);
+        public interface Isa {
+            IsaConstraint isa();
+        }
+
+        public interface Value {
+            ValueConstraint<?> value();
+        }
+
+        public static class Relation extends Conclusion implements Isa {
+
+            private final RelationConstraint relation;
+            private final IsaConstraint isa;
+
+            public Relation(RelationConstraint relation, IsaConstraint isa) {
```

Comment:
actually on second thought its clearer if the constructor takes the entire conjunction, and extracts the constraints in here. This way the assertions can also sit inside the constructor, which feels like a better place?

Reply:
Resolved in conversation: now using static constructors that return optionals to move the validation logic down into the Conclusion classes.